akka-stream-tests/src/test/scala/akka/stream/impl/fusing/ActorGraphInterpreterSpec.scala

```diff
@@ -256,5 +258,65 @@ class ActorGraphInterpreterSpec extends AkkaSpec {
 
     }
 
+    "be able to properly handle case where a stage fails before subscription happens" in assertAllStagesStopped {
+
+      // Fuzzing needs to be off, so that the failure can propagate to the output boundary before the ExposedPublisher
+      // message.
+      val noFuzzMat = ActorMaterializer(ActorMaterializerSettings(system).withFuzzing(false))
+
+      val te = TE("Test failure in preStart")
+
+      val latch = new CountDownLatch(1)
+
+      /*
+       * This is a somewhat tricky test setup. We need the following conditions to be met:
+       *  - the stage should fail its output port before the ExposedPublisher message is processed
+       *  - the enclosing actor (and therefore the stage) should be kept alive until a stray SubscribePending arrives
+       *    that has been enqueued after ExposedPublisher message has been enqueued, but before it has been processed
+       *
+       * To achieve keeping alive the stage for long enough, we use an extra port and instead of failing the stage, we
+       * fail only the output port under test.
+       *
+       * To delay the startup long enough, so both ExposedPublisher and SubscribePending are enqueued, we use an evil
+       * latch to delay the preStart() (which in turn delays the enclosing actor's preStart).
+       *
+       * The test is still somewhat racy but I couldn't find a better way so far.
+       */
+
+      val failyStage = new GraphStage[FlowShape[Int, Int]] {
+        override val shape: FlowShape[Int, Int] = new FlowShape(Inlet[Int]("test.in"), Outlet[Int]("test.out"))
+
+        override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+
+          override def preStart(): Unit = {
+            latch.await()
+            fail(shape.out, te)
+          }
+
+          override def postStop(): Unit = testActor ! "I am dead"
+
+          setHandler(shape.out, GraphStageLogic.IgnoreTerminateOutput) //We fail in preStart anyway
+          setHandler(shape.in, GraphStageLogic.EagerTerminateInput)
+
+        }
+      }
+
+      val downstream = TestSubscriber.probe[Int]()
+      val upstream = TestPublisher.probe[Int]()
+
+      Source.fromPublisher(upstream).via(failyStage).runWith(Sink.fromSubscriber(downstream))
+      latch.countDown()
+
+      // If an NPE would happen due to unset exposedPublisher (see #19338) this would receive the "I am dead" message
+      // do to postStop being executed earlier than expected.
+      expectNoMsg(500.millis)
```

Comment:
what you could do is to use another output port from which we expect a single element that comes from the source, then we’d see the NPE bubbling up there, right?